File: apps/mobile/database/models/Hive.ts
import { Model, type Query } from "@nozbe/watermelondb";
import { text, field, date, children, relation, readonly } from "@nozbe/watermelondb/decorators";
import type { Associations } from "@nozbe/watermelondb/Model";
import type { Relation } from "@nozbe/watermelondb/Model";
import type Apiary from "./Apiary";
import type Queen from "./Queen";
import type Inspection from "./Inspection";
import type Treatment from "./Treatment";
import type Harvest from "./Harvest";
import type Event from "./Event";
import type Task from "./Task";
import type TaskCadence from "./TaskCadence";

export default class Hive extends Model {
  static table = "hives";

  static associations: Associations = {
    apiaries: { type: "belongs_to" as const, key: "apiary_id" },
    queens: { type: "has_many" as const, foreignKey: "hive_id" },
    inspections: { type: "has_many" as const, foreignKey: "hive_id" },
    treatments: { type: "has_many" as const, foreignKey: "hive_id" },
    harvests: { type: "has_many" as const, foreignKey: "hive_id" },
    events: { type: "has_many" as const, foreignKey: "hive_id" },
    tasks: { type: "has_many" as const, foreignKey: "hive_id" },
    task_cadences: { type: "has_many" as const, foreignKey: "hive_id" },
  };

  @text("apiary_id") apiaryId!: string;
  @text("name") name!: string;
  @text("hive_type") hiveType!: string;
  @text("status") status!: string;
  @text("source") source!: string | null;
  @text("installation_date") installationDate!: string | null;
  @text("color") color!: string | null;
  @field("position_order") order!: number | null;
  @text("notes") notes!: string | null;
  @readonly @date("created_at") createdAt!: Date;
  @readonly @date("updated_at") updatedAt!: Date;

  @relation("apiaries", "apiary_id") apiary!: Relation<Apiary>;
  @children("queens") queens!: Query<Queen>;
  @children("inspections") inspections!: Query<Inspection>;
  @children("treatments") treatments!: Query<Treatment>;
  @children("harvests") harvests!: Query<Harvest>;
  @children("events") events!: Query<Event>;
  @children("tasks") tasks!: Query<Task>;
  @children("task_cadences") cadences!: Query<TaskCadence>;
}

File: apps/api/app/models/hive.py
"""Hive model."""

import enum
import uuid
from datetime import date

from sqlalchemy import Date, ForeignKey, Integer, String, Text
from sqlalchemy import Enum as SAEnum
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.models.base import Base


class HiveType(enum.StrEnum):
    LANGSTROTH = "langstroth"
    TOP_BAR = "top_bar"
    WARRE = "warre"
    FLOW = "flow"
    OTHER = "other"


class HiveStatus(enum.StrEnum):
    ACTIVE = "active"
    DEAD = "dead"
    COMBINED = "combined"
    SOLD = "sold"


class HiveSource(enum.StrEnum):
    PACKAGE = "package"
    NUC = "nuc"
    SWARM = "swarm"
    SPLIT = "split"
    PURCHASED = "purchased"


class Hive(Base):
    __tablename__ = "hives"

    apiary_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("apiaries.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    name: Mapped[str] = mapped_column(String(255), nullable=False)
    hive_type: Mapped[HiveType] = mapped_column(
        SAEnum(HiveType, name="hive_type"),
        default=HiveType.LANGSTROTH,
        nullable=False,
    )
    status: Mapped[HiveStatus] = mapped_column(
        SAEnum(HiveStatus, name="hive_status"),
        default=HiveStatus.ACTIVE,
        nullable=False,
    )
    source: Mapped[HiveSource | None] = mapped_column(
        SAEnum(HiveSource, name="hive_source"), nullable=True
    )
    installation_date: Mapped[date | None] = mapped_column(Date, nullable=True)
    color: Mapped[str | None] = mapped_column(String(50), nullable=True)
    order: Mapped[int | None] = mapped_column(Integer, nullable=True)
    notes: Mapped[str | None] = mapped_column(Text, nullable=True)

    # Relationships
    apiary: Mapped["Apiary"] = relationship("Apiary", back_populates="hives")
    queen: Mapped["Queen | None"] = relationship(
        "Queen", back_populates="hive", uselist=False, cascade="all, delete-orphan"
    )
    inspections: Mapped[list["Inspection"]] = relationship(
        "Inspection", back_populates="hive", cascade="all, delete-orphan"
    )
    treatments: Mapped[list["Treatment"]] = relationship(
        "Treatment", back_populates="hive", cascade="all, delete-orphan"
    )
    harvests: Mapped[list["Harvest"]] = relationship(
        "Harvest", back_populates="hive", cascade="all, delete-orphan"
    )
    events: Mapped[list["Event"]] = relationship(
        "Event", back_populates="hive", cascade="all, delete-orphan"
    )
    tasks: Mapped[list["Task"]] = relationship("Task", back_populates="hive")
    cadences: Mapped[list["TaskCadence"]] = relationship(
        "TaskCadence", back_populates="hive", cascade="all, delete-orphan"
    )
